### PR TITLE
Use index file for open-source hub.

### PR DESCRIPTION
**Pull request by r0mant**
This PR updates the open-source hub client to use index file for discovery. The index file is stored in the same bucket on S3. The format that's being used is Helm chart repository index file. Index file contains various metadata about installers stored in the hub which allows tools like `tele ls` show better output. The structure of the hub bucket hasn't changed.

The index file is updated every time a new release is published to the hub, see https://github.com/gravitational/ops/pull/105 for details. To implement this, I've added a new (hidden) command, `gravity app index` that generates a proper index file for our applications.

Closes https://github.com/gravitational/gravity.e/issues/3911.

### CODE REVIEW

File: tool/gravity/cli/helm.go [hunk 1]
```diff
@@ -385,6 +388,40 @@ func appRebuildIndex(env *localenv.LocalEnvironment) error {
...
+		}
+		mergeIndexFile.Merge(indexFile)
+		mergeIndexFile.SortEntries()
+		indexFile = mergeIndexFile
+	}
+	bytes, err := yaml.Marshal(indexFile)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	env.Printf(string(bytes))
```
Comment (a-palchikov) [inline tool/gravity/cli/helm.go]: nit: `env.Print`

File: lib/helm/index_test.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+		cluster1, app1, cluster2, runtime1, app2,
+	})
+	c.Assert(err, check.IsNil)
+	// Nullify timestampts so we can deep compare.
+	for _, versions := range indexFile.Entries {
+		for _, version := range versions {
+			version.Created = time.Time{}
+		}
+	}
+	c.Assert(len(indexFile.Entries), check.Equals, 3)
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/helm/index_test.go]: `indexFile.Entries` is a map though.

File: lib/hub/hub.go [hunk 3]
```diff
@@ -63,6 +65,10 @@ import (
 //               ∟ telekube-5.2.0-linux-x86_64.tar.sha256
 //         ∟ latest: // same as versioned sub-bucket
 //         ∟ stable: // same as versioned sub-bucket
+//
+// The index file, index.yaml, provides information about installers stored
+// in the bucket and is updated every time a new version is published. The
+// index file format is the same as Helm's chart repository index file.
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/hub/hub.go]: Yeah, there's no removing. We've never removed releases as far as I remember.

File: lib/hub/hub.go [hunk 4]
```diff
@@ -257,6 +234,28 @@ func (h *s3Hub) Get(locator loc.Locator) (io.ReadCloser, error) {
...
+func (h *s3Hub) getIndexFile() (*repo.IndexFile, error) {
+	object, err := h.S3.GetObject(&s3.GetObjectInput{
+		Bucket: aws.String(h.Bucket),
+		Key:    aws.String(filepath.Join(h.Prefix, indexFileName)),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer object.Body.Close()
+	bytes, err := ioutil.ReadAll(object.Body)
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/hub/hub.go]: This is yaml, not json and the library we use doesn't have a decoder analog it seems.

File: lib/helm/index_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (a-palchikov) [reply lib/helm/index_test.go]: Even better - `lib/compare/DeepEquals` uses `reflect.DeepEquals`:
```
Map values are deeply equal when all of the following are true: they are both nil or both non-nil, they have the same length, and either they are the same map object or their corresponding keys (matched using Go equality) map to deeply equal values.
```

[the remark this answers is not in the document]
Comment (r0mant) [reply lib/helm/index_test.go]: Huh, cool.

File: lib/hub/hub.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (a-palchikov) [reply lib/hub/hub.go]: `k8s.io/apimachinery/pkg/util/yaml` has a streaming YAML decoder

**Pull request merged**